Source/ProgressionSystemRuntime/Private/Components/PSHUDComponent.cpp

```diff
@@ -0,0 +1,178 @@
+﻿// Copyright (c) Valerii Rotermel & Yevhenii Selivanov
+
+#include "Components/PSHUDComponent.h"
+//---
+#include "Bomber.h"
+#include "Controllers/MyPlayerController.h"
+#include "UtilityLibraries/MyBlueprintFunctionLibrary.h"
+#include "Data/PSDataAsset.h"
+#include "Blueprint/WidgetTree.h"
+#include "UI/MyHUD.h"
+#include "Widgets/PSMenuWidget.h"
+#include "Data/PSTypes.h"
+#include "Data/PSSaveGameData.h"
+#include "Data/PSWorldSubsystem.h"
+#include "Components/Image.h"
+#include "GameFramework/MyGameStateBase.h"
+#include "GameFramework/MyPlayerState.h"
+#include "MyDataTable/MyDataTable.h"
+#include "MyUtilsLibraries/WidgetUtilsLibrary.h"
+
+#include UE_INLINE_GENERATED_CPP_BY_NAME(PSHUDComponent)
+
+// Sets default values for this component's properties
+UPSHUDComponent::UPSHUDComponent()
+{
+	// Set this component to be initialized when the game starts, and to be ticked every frame.  You can turn these features
+	// off to improve performance if you don't need them.
+	PrimaryComponentTick.bCanEverTick = false;
+	PrimaryComponentTick.bStartWithTickEnabled = false;
+}
+
+// Called when the game starts
+void UPSHUDComponent::BeginPlay()
+{
+	Super::BeginPlay();
+
+	AMyHUD* MyHUD = Cast<AMyHUD>(GetOwner());
+	const AMyHUD& HUD = *MyHUD;
+	ProgressionMenuWidgetInternal = HUD.CreateWidgetByClass<UPSMenuWidget>(UPSWorldSubsystem::Get().GetPSDataAsset()->GetProgressionMenuWidget(), true, 1);
```

Comment:
GetPSDataAsset-> could cause silent crash. I understand, crash is designed behavior, we want to crash the game if data asset is null, but not silently.

1. Please create inside UPSDataAsset class the static `Get()` function that returns reference (&) and crashes with `checkf` if PSWorldSubsystem::GetPSDataAsset returns null data asset.
2. Use everywhere in code `UPSDataAsset::Get().XXX` instead of `UPSWorldSubsystem::Get().GetPSDataAsset()->`

See next example: `UNMMDataAsset::Get()`

Reply:
solved. 
So the lesson is: always access the DataAssest directly, even it's initialized in submodule?

Reply:
Own `UPSDataAsset::Get().X` method guarantees you:
- Data Asset is always loaded (even if not, then this getter will load it). Without getter usage, it might be not loaded if called on the beginning as you just had.
- The game will always crash with fatal logged error if asset was not loaded for any reason. Without it, you have to everywhere verify it nullptr on put `checkf` many times